### ISSUE TITLE
RFC: Deny list for CQL partition 

### ISSUE THREAD

**Issue by tzach** [scylladb/scylladb#13934]
Based on Apache Cassandra 4.1 feature
https://issues.apache.org/jira/browse/CASSANDRA-12106
https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-13%3A+Denylisting+partitions

A reads/writes to a given partition may cause issues for whatever reason. It would be useful to have a manual way to block requests to such partitions. 

API should allow to list/add/remove partitions from the Deny list.

The input to the list should work with the output of the existing [toppartitions command](https://docs.scylladb.com/stable/operating-scylla/nodetool-commands/toppartitions.html)

Open question:
- Make this local node operation or cluster wide?
- Use SQL/REST API or both?
- Ability to block only READ / WRITES
- Which Metrics to expose?

**Comment by avikivity** [comment]
We have PER PARTITION RATE LIMIT.

**Comment by dorlaor** [comment]
On Thu, May 18, 2023 at 1:18 PM Avi Kivity ***@***.***> wrote:

> We have PER PARTITION RATE LIMIT.
>

In the general case, per partition limit is much better.
This api is for disasters and for extreme edge cases where you just want to
skip
a partition entirely

**Comment by avikivity** [comment]
Do we have examples where we would have needed it?

**Comment by tzach** [comment]
An example can be a too-large partition.
Limiting the rate limit for *all* parties won't help in this case.

Blocking (denaying) user reading/writing might not be enough.
Other operations, like a compact, stream, and MV build, might fail when trying to read the problematic partition.

**Comment by nyh** [comment]
> Do we have examples where we would have needed it?

The commit where Cassandra added this feature is

https://github.com/apache/cassandra/commit/ab920c30310a8c095ba76b363142b8e74cbf0a0a

They have a document which supposedly explains why this feature was added.

For 100% compatibility with Cassandra we would need to support this feature too (perhaps as a wrapper around our more general rate-limiting option, with a rate limit of 0), but I agree that it's not important to have 100% compatibility with Cassandra when it comes to **configuration**.

**Comment by tzach** [comment]
One use case for Deny list can be too large partition
https://github.com/scylladb/scylladb/issues/13934

**Comment by avikivity** [comment]
Reference loop detected

**Comment by tzach** [comment]
> Reference loop detected

fixed

**Comment by vladzcloudius** [comment]
@nyh note that if you set a rate-limit to 0 you are blocking not just a single partition but any partition in the corresponding table AFAIU.

This proposal is about blacklisting a specific partition.

It can be useful for example if the DevOp detects a partition that causes a lot of trouble to a cluster but he/she has to wait for a confirmation before deleting it.
While he/she waits he can blacklist this partition to eliminate a negative effect it causes.

**Comment by nyh** [comment]
@vladzcloudius, you're right. I was confused by @avikivity's statement about the availability of our **per-partition** rate limit feature. Apparently that feature isn't about rate-limiting of any specific partition, but about rate-limiting each and every partition to one rate limit common to all partitions. So you can say for example that you don't want any individual partition to be hit more than 100 requests per second. This is very different from the feature requested in this issue.
